persistence/sql/multi-tenant.md

```diff
@@ -0,0 +1,81 @@
+---
+title: Multi-tenant support
+summary: Highlights differences in the database schema between the two persisters
+reviewed: 2019-03-06
+component: SqlPersistence
+versions: '[4.5,)'
+related:
+ - nservicebus/pipeline/manipulate-with-behaviors
+ - samples/multi-tenant/propagation
+---
+
+When working in a multi-tenant system, the data for each customer/client (tenant) is stored in an independent database identified by a message header passed along with each message flowing through the system. With SQL Persistence running in multi-tenant mode, [saga data](saga.md) and [outbox data](outbox.md) are both stored in the same database as the tenant data, requiring only one database connection and transaction for the duration of the message handler.
+
+[Timeout data](timeouts.md) and [subscription data](subscriptions.md) do not belong to any specific tenant and are stored in a master database in cases where the message transport does not provide native timeouts or native publish/subscribe capabilities.
+
+## Specifying connections per tenant
+
+If the tenant information is propagated in a single header that does not change, multi-tenancy can be enabled by specifying the header name and a `Func` to create a connection given a tenant id:
+
+snippet: MultiTenantWithHeaderName
+
+In more complex situations, where the tenant id must be calculated by consulting multiple headers, or where a transition from an old header name to a new header name is occurring, a `Func` can be provided that captures the tenant id from the incoming message.
+
+In this example, the header `NewTenantHeaderName` is consulted first, with `OldTenantHeaderName` as a backup.
+
+snippet: MultiTenantWithFunc
+
+NOTE: A null tenant id probably indicates a failure to propagate the tenant id from an older message, rendering the message invalid. However, it is safe to return null in this `Func` because SQL Persistence will throw an exception if a null tenant id is encountered, rather than trying to process a message without a tenant id.
```

Comment:
Thinking out loud: is this always the case? Or is it safer to assume that in a multi-tenant system there might be messages that are not tenant-specific?

Reply:
I would say that an endpoint is either in a multi-tenant mode and can only process multi-tenant messages or in a single tenant mode. I'd rather not mix these in a single endpoint.

Also I'd remove `probably` from the sentence above.

Reply:
Fair enough. One scenario, though, could be: if system is currently a home made multi tenant system as it was the only way to build it. Using a message property to transport the tenant id, there is no way to smoothly transition to a header based tenant id as all in flight messages will fail.

Not sure it’s a scenario we need to support. Thoughts?
-

Reply:
The connection for the outbox is needed way before the body is deserialized so we are only talking here about a scenario where the user does not use the outbox.

In that case the fist time the `ConnectionManager` is used to build the connection is when creating the Synchronized Storage Session right before executing a handler. In that case I think it should be possible to introduce a behavior in the incoming logical stage that gets the tenant ID from the deserialized message body and sticks it into a header of the `IncomingMessage` object.

Reply:
Like it 👍